### PR TITLE
Add Dataframe dataset

### PR DESCRIPTION
**Pull request by utkarsharma2**
# Description
## What is the current behavior?
Currently, the dataframe dataset is not supported.

closes: https://github.com/astronomer/apache-airflow-provider-transfers/issues/18 https://github.com/astronomer/apache-airflow-provider-transfers/issues/21

## What is the new behavior?
Added dataframe dataset.

## Does this introduce a breaking change?
Nope

### Checklist
- [ ] Created tests which fail without the change (if possible)
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: src/universal_transfer_operator/data_providers/base.py [hunk 1]
```diff
@@ -37,33 +34,8 @@ class DataProviders(ABC, Generic[DatasetType]):
...
-        self.transfer_params = transfer_params
-        self.transfer_mode = transfer_mode
-        self.transfer_mapping: set[Location] = set()
-        self.LOAD_DATA_NATIVELY_FROM_SOURCE: dict = {}
-
-    def __repr__(self):
-        return f'{self.__class__.__name__}(conn_id="{self.dataset.conn_id})'
-
-    @property
-    def hook(self) -> BaseHook:
```
Comment (sunank200) [inline src/universal_transfer_operator/data_providers/base.py:53]: Why is this removed from the base dataprovider?

File: src/universal_transfer_operator/datasets/dataframe/base.py [hunk 2]
```diff
@@ -10,12 +12,17 @@ class Dataframe(Dataset):
...
 
-    name: str | None = None
+    name: str = field(default="")
+    dataframe: Any = field(default=None)
+    uri: str = field(init=None)  # type: ignore
 
     # TODO: define the name and namespace for dataframe
+
+    @uri.default
+    def _path_to_dataset_uri(self) -> str:
```
Comment (sunank200) [inline src/universal_transfer_operator/datasets/dataframe/base.py:26]: How do you make sure the path to the dataframe dataset uri is unique?

**Review by sunank200 — CHANGES_REQUESTED** [review]
Does this PR require changes for base dataprovider class? Added few comments

**Review by pankajkoti — COMMENTED** [review]
Are we still planning to work on this? I could not review myself as a reviewer somehow, what could be the cause? It appears in my pending reviews list.

File: src/universal_transfer_operator/data_providers/base.py (hunk 1 above)
Comment (utkarsharma2) [reply src/universal_transfer_operator/data_providers/base.py:53]: Since we are adding dataframe dataset, which will only be in memory construct, now that we are catering to in-memory objects only, having hook, openlineage_dataset_namespace, openlineage_dataset_name, and openlineage_dataset_uri don't make sense to me. What I think should be done is create a base class free of all the destination-specific and have a structure-like. WDYT?


`Dataset`
|__> `volatileDatasets`
          |__> `Dataframe`
|__>  `NonVolatileDatasets 
          |__> `TableDataset`
          |__> `FileDataset`

**Comment by utkarsharma2** [comment]
@sunank200 @phanikumv In addition to this thread - https://astronomer.slack.com/archives/C03868KGF2Q/p1682419076485949. I realized that we should only have an in-memory construct for Dataframe because if you consider transfers as mentioned below, they can be represented as existing transfers.
1. Dataframe(dataset=Table()) to Dataframe(dataset=File()) -- which is same as existing Table() to File()
2. Dataframe(dataset=File()) to Dataframe(dataset=Table()) -- which is same as existing File() to Table()
3. Dataframe(dataset=File()) to Dataframe(dataset=File()) -- which is same as existing File() to File()
4. Dataframe(dataset=Table()) to Dataframe(dataset=Table()) -- which is same as existing File() to File()

The only transfer that makes sense to me are:
1. Dataframe(dataset=Table()) to Dataframe(dataset=pandas()) -- which will return an in-memory dataframe.
2. Dataframe(dataset=pandas()) to Dataframe(dataset=Table())  -- which will transfer an in-memory data frame to a File
same cases for the table to dataframe.

File: src/universal_transfer_operator/data_providers/base.py [hunk 3]
```diff
@@ -75,33 +47,3 @@ def write(self, source_ref: pd.DataFrame | DataStream) -> str:  # type: ignore
         :param source_ref: Stream of data to be loaded into output table or a pandas dataframe.
         """
         raise NotImplementedError
-
-    @property
-    def openlineage_dataset_namespace(self) -> str:
```
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply src/universal_transfer_operator/data_providers/base.py:80]: refer - https://github.com/astronomer/apache-airflow-provider-transfers/pull/63#discussion_r1245268367

File: src/universal_transfer_operator/data_providers/base.py [hunk 4]
```diff
@@ -75,33 +47,3 @@ def write(self, source_ref: pd.DataFrame | DataStream) -> str:  # type: ignore
...
-    @property
-    def openlineage_dataset_name(self) -> str:
-        """
-        Returns the open lineage dataset name as per
-        https://github.com/OpenLineage/OpenLineage/blob/main/spec/Naming.md
-        """
-        raise NotImplementedError
-
-    @property
-    def openlineage_dataset_uri(self) -> str:
```
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply src/universal_transfer_operator/data_providers/base.py:96]: refer - https://github.com/astronomer/apache-airflow-provider-transfers/pull/63#discussion_r1245268367

File: src/universal_transfer_operator/datasets/dataframe/base.py (hunk 2 above)
Comment (utkarsharma2) [reply src/universal_transfer_operator/datasets/dataframe/base.py:26]: I don't have a good option for this, hence kept it open. A few options that I considered are
1. Hash of that data, but that can change.
2. Memory address. But this needs to be reconstructed for different dag which might be running on different machines.

I'm open to suggestions here.

Comment (sunank200) [reply src/universal_transfer_operator/datasets/dataframe/base.py:26]: I like approach one.

File: src/universal_transfer_operator/data_providers/database/base.py [hunk 5]
```diff
@@ -650,11 +651,13 @@ def load_pandas_dataframe_to_table(
         )
 
     @staticmethod
-    def _assert_not_empty_df(df):
+    def _assert_not_empty_df(df: pd.DataFrame):
```
Comment (Lee-W) [inline src/universal_transfer_operator/data_providers/database/base.py:654]: nitpick: missing return type

File: src/universal_transfer_operator/data_providers/database/base.py [hunk 6]
```diff
@@ -650,11 +651,13 @@ def load_pandas_dataframe_to_table(
...
         """Raise error if dataframe empty
 
         param df: A dataframe
         """
+        # if isinstance(df, PandasdataframeDataProvider):
+        #     df = df.dataset
```
Comment (Lee-W) [inline src/universal_transfer_operator/data_providers/database/base.py:660]:
```suggestion
```

File: src/universal_transfer_operator/data_providers/database/base.py [hunk 7]
```diff
@@ -381,6 +378,10 @@ def create_table_using_schema_autodetection(
             source_dataframe = file.export_to_dataframe(nrows=LOAD_TABLE_AUTODETECT_ROWS_COUNT)
 
         db = SQLDatabase(engine=self.sqlalchemy_engine)
+
+        # if isinstance(source_dataframe, PandasdataframeDataProvider):
+        #     source_dataframe = source_dataframe.dataset
```
Comment (Lee-W) [inline src/universal_transfer_operator/data_providers/database/base.py:383]:
```suggestion
```

File: src/universal_transfer_operator/data_providers/database/base.py [hunk 8]
```diff
@@ -743,5 +746,21 @@ def export_table_to_pandas_dataframe(self) -> pd.DataFrame:
...
 
         sqla_table = self.get_sqla_table(self.dataset)
-        df = pd.read_sql(sql=sqla_table.select(), con=self.sqlalchemy_engine)
-        return PandasDataframe.from_pandas_df(df)
+        # df = pd.read_sql(sql=sqla_table.select(), con=self.sqlalchemy_engine)
+        # from universal_transfer_operator.data_providers.dataframe.Pandasdataframe import (
+        #     PandasdataframeDataProvider,
+        # )
+        #
+        # return PandasdataframeDataProvider.from_pandas_df(df)
```
Comment (Lee-W) [inline src/universal_transfer_operator/data_providers/database/base.py:754]:
```suggestion
```

File: src/universal_transfer_operator/data_providers/database/base.py [hunk 9]
```diff
@@ -743,5 +746,21 @@ def export_table_to_pandas_dataframe(self) -> pd.DataFrame:
...
+    def is_native_path_available(
+        self,
+        source_dataset: Table,  # skipcq PYL-W0613, PYL-R0201
+    ) -> bool:
+        """
+        Check if there is an optimised path for source to destination.
+
+        :param source_dataset: Dataframe from which we need to transfer data
+        """
+        return False
```
Comment (Lee-W) [inline src/universal_transfer_operator/data_providers/database/base.py:766]: Why do we always return False?

File: src/universal_transfer_operator/data_providers/dataframe/Pandasdataframe.py [hunk 10]
```diff
@@ -49,11 +48,37 @@ def convert_dataframe_to_file(df: pd.DataFrame) -> File:
...
-    """Pandas-compatible dataframe class that can be serialized and deserialized into XCom by Airflow 2.5"""
-
+class PandasdataframeDataProvider(DataframeProvider):
     version: ClassVar[int] = 1
 
+    def read(self) -> Iterator[pd.DataFrame]:
+        """Read from dataframe dataset and write to local reference locations or dataframes"""
+        yield self.dataset.dataframe
+
+    def write(self, source_ref: pd.DataFrame | DataStream):
```
Comment (Lee-W) [inline src/universal_transfer_operator/data_providers/dataframe/Pandasdataframe.py:58]: nitpick: missing return type

File: src/universal_transfer_operator/datasets/file/types/json.py [hunk 11]
```diff
@@ -32,11 +28,13 @@ def export_to_dataframe(
...
-            df=df, columns_names_capitalization=columns_names_capitalization
-        )
-        return PandasDataframe.from_pandas_df(df)
+        return pd.read_json(stream, **kwargs_copy)
+        # df = pd.read_json(stream, **kwargs_copy)
+        # from universal_transfer_operator.data_providers.dataframe.Pandasdataframe import (
+        #     PandasdataframeDataProvider,
+        # )
+        #
+        # return PandasdataframeDataProvider.from_pandas_df(df)
```
Comment (Lee-W) [inline src/universal_transfer_operator/datasets/file/types/json.py:37]:
```suggestion

```

File: src/universal_transfer_operator/datasets/file/types/parquet.py [hunk 12]
```diff
@@ -29,11 +25,13 @@ def export_to_dataframe(self, stream, columns_names_capitalization="original", *
...
-            df=df, columns_names_capitalization=columns_names_capitalization
-        )
-        return PandasDataframe.from_pandas_df(df)
+        return pd.read_parquet(byte_io_buffer, **kwargs_copy)
+        # df = pd.read_parquet(byte_io_buffer, **kwargs_copy)
+        # from universal_transfer_operator.data_providers.dataframe.Pandasdataframe import (
+        #     PandasdataframeDataProvider,
+        # )
+        #
+        # return PandasdataframeDataProvider.from_pandas_df(df)
```
Comment (Lee-W) [inline src/universal_transfer_operator/datasets/file/types/parquet.py:34]:
```suggestion

```

File: src/universal_transfer_operator/data_providers/base.py [hunk 13]
```diff
@@ -37,33 +34,8 @@ class DataProviders(ABC, Generic[DatasetType]):
     def __init__(
         self,
         dataset: DatasetType,
-        transfer_mode,
```
Comment (sunank200) [inline src/universal_transfer_operator/data_providers/base.py:40]: is transfer mode not required as part of this ABC?

File: src/universal_transfer_operator/data_providers/database/base.py [hunk 14]
```diff
@@ -381,6 +378,10 @@ def create_table_using_schema_autodetection(
             source_dataframe = file.export_to_dataframe(nrows=LOAD_TABLE_AUTODETECT_ROWS_COUNT)
 
         db = SQLDatabase(engine=self.sqlalchemy_engine)
+
+        # if isinstance(source_dataframe, PandasdataframeDataProvider):
```
[the remark this answers is not in the document]
Comment (sunank200) [reply src/universal_transfer_operator/data_providers/database/base.py:382]: +1

File: src/universal_transfer_operator/data_providers/database/base.py (hunk 6 above)
Comment (sunank200) [reply src/universal_transfer_operator/data_providers/database/base.py:660]: +1

File: src/universal_transfer_operator/data_providers/database/base.py (hunk 8 above)
Comment (sunank200) [reply src/universal_transfer_operator/data_providers/database/base.py:754]: +1

File: src/universal_transfer_operator/data_providers/database/base.py (hunk 9 above)
Comment (sunank200) [reply src/universal_transfer_operator/data_providers/database/base.py:766]: By default, these data providers don't have native transfers. If any specific data provider has native transfer enabled, it overrides this.

File: src/universal_transfer_operator/data_providers/dataframe/Pandasdataframe.py (hunk 10 above)
Comment (sunank200) [reply src/universal_transfer_operator/data_providers/dataframe/Pandasdataframe.py:58]: +1

File: src/universal_transfer_operator/data_providers/dataframe/base.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+from universal_transfer_operator.datasets.dataframe.base import Dataframe
+
+
+class DataframeProvider(DataProviders[Dataframe]):
+    """Base class to import dataframe implementation. We intend to support different implementation of dataframes."""
+
+    def __init__(
+        self,
+        dataset: Dataframe,
+    ):
```
[the remark this answers is not in the document]
Comment (sunank200) [reply src/universal_transfer_operator/data_providers/dataframe/base.py:16]: +1

File: src/universal_transfer_operator/data_providers/dataframe/base.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+    def read(self) -> Iterator[Dataframe]:
+        """Read from dataframe dataset and write to local reference locations or dataframes"""
+        raise NotImplementedError
+
+    def write(self, source_ref: pd.DataFrame | DataStream) -> str:  # type: ignore
+        """Write the data to the dataframe dataset or filesystem dataset"""
+        raise NotImplementedError
+
+    def serialize(self):
+        # Store in the metadata DB if Dataframe
```
[the remark this answers is not in the document]
Comment (sunank200) [reply src/universal_transfer_operator/data_providers/dataframe/base.py:29]: +1

File: src/universal_transfer_operator/data_providers/dataframe/base.py [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+        raise NotImplementedError
+
+    @staticmethod
+    def deserialize(data: dict, version: int):
+        # Extract from metadata DB
```
[the remark this answers is not in the document]
Comment (sunank200) [reply src/universal_transfer_operator/data_providers/dataframe/base.py:34]: +1

File: src/universal_transfer_operator/data_providers/filesystem/base.py [hunk 18]
```diff
@@ -225,18 +223,20 @@ def size(self) -> int:
...
+    def is_native_path_available(
+        self,
+        source_dataset: File,  # skipcq PYL-W0613, PYL-R0201
+    ) -> bool:
+        """
+        Check if there is an optimised path for source to destination.
+
+        :param source_dataset: File from which we need to transfer data
+        """
+        return False
```
[the remark this answers is not in the document]
Comment (sunank200) [reply src/universal_transfer_operator/data_providers/filesystem/base.py:242]: Answered in above thread

File: src/universal_transfer_operator/datasets/file/types/csv.py [hunk 19]
```diff
@@ -26,11 +22,13 @@ def export_to_dataframe(
...
-            df=df, columns_names_capitalization=columns_names_capitalization
-        )
-        return PandasDataframe.from_pandas_df(df)
+        return pd.read_csv(stream, **kwargs)
+        # df = pd.read_csv(stream, **kwargs)
+        # from universal_transfer_operator.data_providers.dataframe.Pandasdataframe import (
+        #     PandasdataframeDataProvider,
+        # )
+        #
+        # return PandasdataframeDataProvider.from_pandas_df(df)
```
[the remark this answers is not in the document]
Comment (sunank200) [reply src/universal_transfer_operator/datasets/file/types/csv.py:31]: +1

File: src/universal_transfer_operator/datasets/file/types/json.py (hunk 11 above)
Comment (sunank200) [reply src/universal_transfer_operator/datasets/file/types/json.py:37]: +1

**Review by sunank200 — COMMENTED** [review]
Can we add an integration test for this? How are we testing end to end transfer between various providers and their combinations?

File: src/universal_transfer_operator/datasets/file/types/ndjson.py [hunk 20]
```diff
@@ -24,11 +21,13 @@ def export_to_dataframe(self, stream, columns_names_capitalization="original", *
         :param columns_names_capitalization: determines whether to convert all columns to lowercase/uppercase
             in the resulting dataframe
         """
-        df = NDJsonFileTypes.flatten(self.normalize_config, stream, **kwargs)
-        df = convert_columns_names_capitalization(
-            df=df, columns_names_capitalization=columns_names_capitalization
-        )
-        return PandasDataframe.from_pandas_df(df)
+        return NDJsonFileTypes.flatten(self.normalize_config, stream, **kwargs)
```
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply src/universal_transfer_operator/datasets/file/types/ndjson.py:24]: it does, `NDJsonFileTypes.flatten` returns dataframe

File: src/universal_transfer_operator/datasets/dataframe/base.py (hunk 2 above)
Comment (utkarsharma2) [reply src/universal_transfer_operator/datasets/dataframe/base.py:26]: @sunank200  I don't think we have a URI for something that only exists in memory because of the reason listed about. That's why users can create a unique string for themselves or we can generate that for them. But If we have a better solution we can implement that.

File: src/universal_transfer_operator/data_providers/__init__.py [hunk 21]
```diff
@@ -23,31 +24,36 @@
...
         transfer_params = TransferIntegrationOptions()
-    data_provider: DataProviders = class_ref(
-        dataset=dataset,
-        transfer_params=transfer_params,
-        transfer_mode=transfer_mode,
-    )
+
+    class_ref: type[DataProviders] = get_dataprovider_class(dataset=dataset)
+    if isinstance(dataset, (Table, File)):
+        data_provider = class_ref(  # type: ignore
```
[the remark this answers is not in the document]
Comment (utkarsharma2) [reply src/universal_transfer_operator/data_providers/__init__.py:41]: MyPy is not respecting the type of dataset here. I'll check how to do it in and address this in a separate PR.

File: src/universal_transfer_operator/data_providers/dataframe/Pandasdataframe.py [hunk 22]
```diff
@@ -49,11 +48,37 @@ def convert_dataframe_to_file(df: pd.DataFrame) -> File:
...
+        else:
+            return PandasdataframeDataProvider(
+                dataset=Dataframe(
+                    dataframe=source_ref.actual_file.type.export_to_dataframe(
+                        stream=source_ref.remote_obj_buffer
+                    )
+                )
+            )
+
+    def equals(self, other: PandasdataframeDataProvider) -> bool:
```
Comment (sunank200) [inline src/universal_transfer_operator/data_providers/dataframe/Pandasdataframe.py:71]: Why do we need this method? Where is this used?

File: src/universal_transfer_operator/universal_transfer_operator.py [hunk 23]
```diff
@@ -65,7 +65,7 @@ def execute(self, context: Context) -> Any:  # skipcq: PYL-W0613
             transfer_params=self.transfer_params,
             transfer_mode=self.transfer_mode,
         )
-        native_path_available = destination_dataprovider.is_native_path_available(
+        native_path_available = destination_dataprovider.is_native_path_available(  # type: ignore
```
Comment (sunank200) [inline src/universal_transfer_operator/universal_transfer_operator.py:68]: Why type ignore here?

**Review by sunank200 — APPROVED** [review]
Overall LGTM

File: src/universal_transfer_operator/data_providers/dataframe/Pandasdataframe.py (hunk 22 above)
Comment (utkarsharma2) [reply src/universal_transfer_operator/data_providers/dataframe/Pandasdataframe.py:71]: It's used in tests when comparing outputs.

File: src/universal_transfer_operator/universal_transfer_operator.py (hunk 23 above)
Comment (utkarsharma2) [reply src/universal_transfer_operator/universal_transfer_operator.py:68]: I'll check that in a separate PR.
